### PR TITLE
Adapt three unit tests with newer compiler/libraries

### PR DESCRIPTION
**Pull request by neilmayhew**
This fixes three tests that fail with relatively recent tools and libraries:

The tests are:

* `spatial_db_test`
* `table_test`
* `db_universal_compaction_test`

I'm using:

* `gcc` 7.3.0
* `glibc` 2.27
* `snappy` 1.1.7
* `gflags` 2.2.1
* `zlib` 1.2.11
* `bzip2` 1.0.6.0.1
* `lz4` 1.8.2
* `jemalloc` 5.0.1

The versions used in the Travis environment (which is two Ubuntu LTS versions behind the current one and doesn't use `lz4` or `jemalloc`) don't seem to have a problem. However, to be safe, I verified that these tests pass with and without my changes in a trusty Docker container without `lz4` and `jemalloc`.

However, I do get an unrelated set of other failures when using a trusty Docker container that uses `lz4` and `jemalloc`:

```
db/db_universal_compaction_test.cc:506: Failure
Value of: num + 1
  Actual: 3
Expected: NumSortedRuns(1)
Which is: 4
[  FAILED  ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/0, where GetParam() = (1, false) (1189 ms)
[ RUN      ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/1
db/db_universal_compaction_test.cc:506: Failure
Value of: num + 1
  Actual: 3
Expected: NumSortedRuns(1)
Which is: 4
[  FAILED  ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/1, where GetParam() = (1, true) (1246 ms)
[ RUN      ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/2
db/db_universal_compaction_test.cc:506: Failure
Value of: num + 1
  Actual: 3
Expected: NumSortedRuns(1)
Which is: 4
[  FAILED  ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/2, where GetParam() = (3, false) (1237 ms)
[ RUN      ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/3
db/db_universal_compaction_test.cc:506: Failure
Value of: num + 1
  Actual: 3
Expected: NumSortedRuns(1)
Which is: 4
[  FAILED  ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/3, where GetParam() = (3, true) (1195 ms)
[ RUN      ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/4
db/db_universal_compaction_test.cc:506: Failure
Value of: num + 1
  Actual: 3
Expected: NumSortedRuns(1)
Which is: 4
[  FAILED  ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/4, where GetParam() = (5, false) (1161 ms)
[ RUN      ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/5
db/db_universal_compaction_test.cc:506: Failure
Value of: num + 1
  Actual: 3
Expected: NumSortedRuns(1)
Which is: 4
[  FAILED  ] UniversalCompactionNumLevels/DBTestUniversalCompaction.DynamicUniversalCompactionReadAmplification/5, where GetParam() = (5, true) (1229 ms)
```

I haven't attempted to fix these since I'm not using trusty and Travis doesn't use `lz4` and `jemalloc`. However, the final commit in this PR does at least fix the compilation errors that occur when using trusty's version of `lz4`.

### CODE REVIEW

**Comment by maysamyabandeh** [comment]
We get this failure in our test infra:
```
[ RUN ] SpatialDBTest.FeatureSetSerializeTest
internal_repo_rocksdb/repo/utilities/spatialdb/spatial_db_test.cc:99: Failure
Value of: !deserialized.Deserialize(serialized)
Actual: false
Expected: true
```

**Comment by neilmayhew** [comment]
@maysamyabandeh It seems to depend on the exact infra that's used, which is weird. In my infra, it succeeds in deserializing the string when 4 chars are removed from the end, which causes the test to fail (because it expects deserialization to fail). On Travis, deserialization fails with both 3 and 4 chars removed, so the test succeeds with and without my change.

I stepped through the code in my infra and I can see that the last 4 bytes are a single unit of deserialization, ie one of the (k,v) pairs. So it makes sense that deserialization would succeed with -4 and fail with -3. However, I haven't been able to step through a -3 case that succeeds, because I can't reproduce that. If you could tell me a little more about the dependencies in your infra, I could try to set up a similar scenario and see if I can reproduce it.

Bottom line, though, removing an arbitrary number bytes from the end of a serialized string seems like a fragile way to test. Is there a more robust way to force a deserialization failure? Perhaps passing a one-byte string containing just the first byte?

**Comment by maysamyabandeh** [comment]
@neilmayhew many unit tests would benefit from a rewrite and I guess this is one of them. I do not have context on this test and cannot tell off the top of my head what is the proper way of rewriting it. If you know a more robust way of doing this test feel free to go ahead and update it.

**Comment by neilmayhew** [comment]
I figured it out. The order of keys in the serialized string in a result of hashing, and some units can be 3 bytes. So in your env the last entry is probably 3 bytes, and in mine it's 4. Using 1 byte is guaranteed to produce a corrupt string, as I explain in the commit message.
